File: 백준/Silver/17413. 단어 뒤집기 2/단어 뒤집기 2.py
import sys
input = sys.stdin.readline

S = list(map(str, input().rstrip()))
res = ''
tf = False
cnt = 0
stack = []

for i in S:
    if i == '<':
        tf = True
        for _ in range(len(stack)):
            res += stack.pop()

    stack.append(i)

    if i == '>':
        tf = False
        for _ in range(len(stack)):
            res += stack.pop(0)
        
    if i == ' ' and tf == False:
        for s in range(len(stack)):
            if s == 0:
                stack.pop()
                continue
            res += stack.pop()
        res += ' '

if stack:
    for _ in range(len(stack)):
        res += stack.pop()
print(res)

File: 백준/Silver/17413. 단어 뒤집기 2/단어 뒤집기 2.java
import java.util.Scanner;
import java.util.Stack;

class Main {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);

        char[] S = input.nextLine().toCharArray();
        Stack<Character> word = new Stack<>();
        boolean flag = false;

        for (int i=0; i<S.length; i++) {
            if (S[i] == '<') {
                while (!word.isEmpty()) {
                    System.out.print(word.pop());
                }
                flag = true;
            } else if (S[i] == '>') {
                System.out.print('>');
                flag = false;
                continue;
            }

            if (flag) {
                System.out.print(S[i]);
            }
            else {
                word.push(S[i]);
                if (S[i] == ' ' || i == S.length-1) {
                    if (S[i] == ' ') word.pop();
                    while (!word.isEmpty()) {
                        System.out.print(word.pop());
                    }
                    if (i != S.length-1) {
                        System.out.print(' ');
                    }
                }
            }
        }
    }
}